### PR TITLE
Correct timeStep in case of missing values

### PR DESCRIPTION
**Pull request by benrubson**
Hello,

This PR correctly calculates timeStep in case of series with missing values.
Due to missing values we could have different timeStep all along the series, so we have to compute the minimum timeStep found for each series.

Thank you 👍 

Ben

### CODE REVIEW

**Comment by benrubson** [comment]
Of course this is related / follows #6287.

**Comment by torkelo** [comment]
not sure this is required, most TSDB's can group by buckets, and create empty buckets etc.

**Comment by torkelo** [comment]
And having different timeSteps between data points and expecting bar width to be fine seems unreasonable

**Comment by benrubson** [comment]
I have the case where for example 0 values are deduplicated (to save bandwith, storage...).
Note that values != 0 could also be deduplicated, this is one of the features of [OpenTSDB/tcollector](https://github.com/OpenTSDB/tcollector).

Then, an example of such a series without this PR :



And with the PR :



Ben

**Comment by torkelo** [comment]
but the deduplication should then be re-added during query time, so there are equal distance between points, this is done by Graphite, InfluxDB, Elasticsearch etc so you get null points, otherwise Grafana will not know of missing data points

**Comment by benrubson** [comment]
Mmmh yes you're right.
It is also done by OpenTSDB, but not with its ZimSum & None aggregators.
My example uses ZimSum aggregator with no fill policy (none).
Perhaps you're scary by an additional calculation in the datapoints loop ?

_Note that #6287 can stack series with null values but also with missing values (such as series using zimsum)._

**Comment by benrubson** [comment]
Many thanks Torkel 👍

File: public/app/core/time_series2.ts [hunk 1]
```diff
@@ -115,6 +115,15 @@ export default class TimeSeries {
...
 
+      // Due to missing values we could have different timeStep all along the series
+      // so we have to find the minimum one (could occur with aggregators such as ZimSum)
+      if (i>0) {
+        var previousTime = this.datapoints[i-1][1];
+        if (!this.stats.timeStep || currentTime - previousTime < this.stats.timeStep) {
+          this.stats.timeStep = currentTime - previousTime;
+        }
+      }
+
```
Comment (torkelo) [inline public/app/core/time_series2.ts]: made some slight refactoring of this here, https://github.com/grafana/grafana/commit/71eb0f3278b63579efe033b058d2363a3df76add

Comment (benrubson) [inline public/app/core/time_series2.ts]: Yep, looks better, perfect 👍

**Pull request merged**